config/src/test/scala/com/typesafe/config/impl/ConfigDocumentTest.scala

```diff
@@ -0,0 +1,257 @@
+                },
+                "c": [1, 2, 3, {"a": "b"}, [1,2,3]]
+              }
+             }"""
+        configDocumentReplaceConfTest(origText, finalText,
+            "this is a concatenation 123 456 {a:b} [1,2,3] {a: this is another 123 concatenation null true}", "h.b.a")
+    }
+
+    @Test
+    def configDocumentSetNewValueBraceRoot {
+        val origText = "{\n\t\"a\":\"b\",\n\t\"c\":\"d\"\n}"
+        val finalTextConf = "{\n\t\"a\":\"b\",\n\t\"c\":\"d\"\n\n\"e\" : \"f\"\n}"
+        val finalTextJson = "{\n\t\"a\":\"b\",\n\t\"c\":\"d\",\n\n\"e\" : \"f\"\n}"
+        configDocumentReplaceConfTest(origText, finalTextConf, "\"f\"", "\"e\"")
+        configDocumentReplaceJsonTest(origText, finalTextJson, "\"f\"", "\"e\"")
+    }
+
+    @Test
+    def configDocumentSetNewValueNoBraces {
+        val origText = "\"a\":\"b\",\n\"c\":\"d\"\n"
+        val finalText = "\"a\":\"b\",\n\"c\":\"d\"\n\n\"e\" : \"f\"\n"
+        configDocumentReplaceConfTest(origText, finalText, "\"f\"", "\"e\"")
+    }
+
+    @Test
+    def configDocumentSetNewValueMultiLevelConf {
+        val origText = "a:b\nc:d"
+        val finalText = "a:b\nc:d\ne : {\nf : {\ng : 12\n}\n}\n"
+        configDocumentReplaceConfTest(origText, finalText, "12", "e.f.g")
+    }
+
+    @Test
+    def configDocumentSetNewValueMultiLevelJson {
+        val origText = "{\"a\":\"b\",\n\"c\":\"d\"}"
+        val finalText = "{\"a\":\"b\",\n\"c\":\"d\",\n\"e\" : {\n\"f\" : {\n\"g\" : 12\n}\n}\n}"
+        configDocumentReplaceJsonTest(origText, finalText, "12", "e.f.g")
+    }
+
+    @Test
+    def configDocumentSetNewConfigValue {
+        val origText = "{\"a\": \"b\"}"
+        val finalText = "{\"a\": 12}"
+        val configDocHOCON = ConfigDocumentFactory.parseString(origText)
+        val configDocJSON = ConfigDocumentFactory.parseString(origText, ConfigParseOptions.defaults.setSyntax(ConfigSyntax.JSON))
+        val newValue = ConfigValueFactory.fromAnyRef(12)
+        assertEquals(origText, configDocHOCON.render())
+        assertEquals(origText, configDocJSON.render())
+        assertEquals(finalText, configDocHOCON.setValue("a", newValue).render())
+        assertEquals(finalText, configDocJSON.setValue("a", newValue).render())
+    }
+
+    @Test
+    def configDocumentArrayReplaceFailure {
+        // Attempting a replace on a ConfigDocument parsed from an array throws an error
+        val origText = "[1, 2, 3, 4, 5]"
+        val document = ConfigDocumentFactory.parseString(origText)
+        var exceptionThrown = false
+        try {
+            document.setValue("a", "1")
+        } catch {
+            case e: Exception =>
+                exceptionThrown = true
+                assertTrue(e.isInstanceOf[ConfigException])
+                assertTrue(e.getMessage.contains("ConfigDocument had an array at the root level"))
+        }
+        assertTrue(exceptionThrown)
+    }
```

Comment:
for future reference, you can write this as `val e = intercept[ConfigException] { document.setValue("a", "1") }` which asserts that the exception of that type is in fact thrown, then you can check the message.